src/platform_impl/android/mod.rs

```diff
@@ -216,383 +189,593 @@ fn ndk_keycode_to_virtualkeycode(keycode: Keycode) -> Option<event::VirtualKeyCo
     }
 }
 
-fn poll(poll: Poll) -> Option<EventSource> {
-    match poll {
-        Poll::Event { ident, .. } => match ident {
-            ndk_glue::NDK_GLUE_LOOPER_EVENT_PIPE_IDENT => Some(EventSource::Callback),
-            ndk_glue::NDK_GLUE_LOOPER_INPUT_QUEUE_IDENT => Some(EventSource::InputQueue),
-            _ => unreachable!(),
-        },
-        Poll::Timeout => None,
-        Poll::Wake => Some(
-            INTERNAL_EVENT
-                .write()
-                .unwrap()
-                .take()
-                .map_or(EventSource::User, EventSource::Internal),
-        ),
-        Poll::Callback => unreachable!(),
+struct PeekableReceiver<T> {
+    recv: mpsc::Receiver<T>,
+    first: Option<T>,
+}
+
+impl<T> PeekableReceiver<T> {
+    pub fn from_recv(recv: mpsc::Receiver<T>) -> Self {
+        Self { recv, first: None }
+    }
+    pub fn has_incoming(&mut self) -> bool {
+        if self.first.is_some() {
+            return true;
+        }
+        match self.recv.try_recv() {
+            Ok(v) => {
+                self.first = Some(v);
+                true
+            }
+            Err(mpsc::TryRecvError::Empty) => false,
+            Err(mpsc::TryRecvError::Disconnected) => {
+                warn!("Channel was disconnected when checking incoming");
+                false
+            }
+        }
+    }
+    pub fn try_recv(&mut self) -> Result<T, mpsc::TryRecvError> {
+        if let Some(first) = self.first.take() {
+            return Ok(first);
+        }
+        self.recv.try_recv()
+    }
+}
+
+#[derive(Clone)]
+struct SharedFlagSetter {
```

Comment:
minor: imo this could be just merged into the `SharedFlag`

Reply:
I suppose I'll mark this one as resolved for now, since it was only noted as 'minor'. At the time of writing it was very intentional from my pov to decouple getting/setting at the type level because conceptually some code is only supposed to have the ability to set the flag and not clear it - so this uses the type system to effectively document the intent and enforce that constraint.

As noted we could combine them if we really figure it's overkill but I also had in mind that a solution like this could be used in other backends too later and in that case it'd maybe be better to keep the more explicit design - and it may be considered less overkill as a shared utility (in case the intention to limit the ability to clear the flag isn't always immediately apparent).